core/app/[locale]/(default)/product/[slug]/_actions/wishlist-action.ts

```diff
@@ -0,0 +1,301 @@
+        result {
+          entityId
+        }
+      }
+    }
+  }
+`);
+
+const menuItemSchema = z.object({
+  action: z.enum(['add', 'remove']),
+  wishlistId: z.number().nonnegative(),
+  wishlistItemId: z.number().optional(),
+  redirectTo: z.string(),
+});
+
+const menuItemParser = z.string().transform((str, ctx) => {
+  try {
+    return menuItemSchema.parse(JSON.parse(str));
+  } catch {
+    ctx.addIssue({ code: 'custom', message: 'Invalid menuItem payload.' });
+
+    return z.NEVER;
+  }
+});
+
+const schema = z.object({
+  productId: z.number().nonnegative().min(1),
+  selectedSku: z.string(),
+  menuItem: menuItemParser,
+});
+
+interface WishlistAddMutationVariables {
+  wishlistId: number;
+  productId: number;
+  variantId?: number;
+}
+
+interface WishlistRemoveMutationVariables {
+  wishlistId: number;
+  wishlistItemId: number;
+}
+
+const getVariantIdFromSku = cache(
+  async (productId: number, sku: string, customerAccessToken?: string) => {
+    const { data } = await client.fetch({
+      document: VariantIdFromSkuQuery,
+      variables: { productId, sku },
+      customerAccessToken,
+      fetchOptions: { next: { revalidate } },
+    });
+
+    return data.site.product?.variants.edges?.[0]?.node.entityId ?? undefined;
+  },
+);
+
+async function addToDefaultWishlist(
+  customerAccessToken: string,
+  wishlistName: string,
+  productId: number,
+  variantId?: number,
+) {
+  const { data } = await client.fetch({
+    document: CreateWishlistMutation,
+    variables: {
+      input: {
+        name: wishlistName,
+        isPublic: false,
+        items: [{ productEntityId: productId, variantEntityId: variantId }],
+      },
+    },
+    customerAccessToken,
+    fetchOptions: { cache: 'no-store' },
+  });
+
+  if (!data.wishlist.createWishlist?.result) {
+    throw new Error('Failed to add item to default wishlist. Response was empty.');
+  }
+}
+
+async function addToWishlist(customerAccessToken: string, variables: WishlistAddMutationVariables) {
+  const { data } = await client.fetch({
+    document: AddToWishlistMutation,
+    variables,
+    customerAccessToken,
+    fetchOptions: { cache: 'no-store' },
+  });
+
+  if (!data.wishlist.addWishlistItems?.result) {
+    throw new Error('Failed to add item to wishlist. Response was empty.');
+  }
+}
+
+async function removeFromWishlist(
+  customerAccessToken: string,
+  variables: WishlistRemoveMutationVariables,
+) {
+  const { data } = await client.fetch({
+    document: DeleteWishlistItemsMutation,
+    variables,
+    customerAccessToken,
+    fetchOptions: { cache: 'no-store' },
+  });
+
+  if (!data.wishlist.deleteWishlistItems?.result) {
+    throw new Error('Failed to add item to wishlist. Response was empty.');
+  }
+}
+
+function getLoginRedirect(redirectTo: string) {
+  const loginParams = new URLSearchParams({ redirectTo });
+
+  return `/login?${loginParams.toString()}`;
+}
```

Comment:
For any sort of `redirectTo`, on the Catalyst side we can only accept relative pathnames to be passed. `URLSearchParams` I think only escapes the param, but we need to pass it into `new URL` and get the pathname (and maybe search params) only. It might make sense to do it right here.

Reply:
@chanceaclark It looks like we're already enforcing this pattern on the login page:
https://github.com/bigcommerce/catalyst/commit/a739f01d0b91c549dae15a1fc350a388c98bbe56#diff-2b5f4d2b2e234bc3346764b7e78954e330e60912ad21040b916b79df65d1fbb7R37-R39

Are we wanting to do this pattern everywhere that we're constructing something to pass to the redirectTo param? I think ultimately it should be fine here since we're explicitly setting a relative path on the formdata that gets passed into this function

Reply:
We should be protecting against this at the boundary. The page is fine, but if someone else consumes this server action (which would be rare), they could bypass our protections.